app/models/user.rb

```diff
@@ -512,6 +515,54 @@ def verify
     save
   end
 
+  # Create a Rails native token (used for authentication in GraphQL API)
+  def create_graphql_token
+    raise("User not verified") unless verified
+
+    token_hash = { user_id: id, in_admin_mode: false }
+    User.encrypted_token(token_hash)
+  end
+
+  def create_admin_token
+    # if they're not an admin, just return false and keep the token
+    token_hash = { user_id: id, in_admin_mode: admin }
```

Comment:
Note that admins are rarely in admin mode.  Admins are the only users who can enter admin mode.  But there's no reason to store whether the user is an admin in the token, since it's in the user.  But whether or not the admin has turned on admin mode is a state, and should be either stored in the session (I forget whether GraphQL uses a session), or in the token.  Very little is kept in the session generally, so I see no reason we can't put it all in the token, personally, and this has the added advantage of keeping GraphQL state-less.

Reply:
I think i'm finally one step ahead of you (it is, and will remain rare, let me enjoy!) This is correctly stored in the token - there's no session. The way it works:

**Login** - `create_graphql_token`. No one's in admin mode at login, `in_admin_mode: false`

**An admin user turns admin mode on** - calls `create_admin_token`. `in_admin_mode: true`

**A non admin tries to turn admin mode on, somehow** - `create_admin_token` simply silently keeps admin mode off. `in_admin_mode: false`

**An admin user turns admin mode off** - calls `create_graphql_token` which again sets `in_admin_mode: false`

That's why it's `in_admin_mode: admin` - it's pulling directly from the user's admin prop.